core-bundle/src/Routing/RouteProvider.php

```diff
@@ -0,0 +1,472 @@
+        $table = $this->pageAdapter->getTable();
+
+        if ($this->configAdapter->get('addLanguageToUrl')) {
+            $path = '/{_locale}'.$path;
+            $requirements['_locale'] = $page->rootLanguage;
+        }
+
+        $routes[$table.'.'.$page->id] = new Route(
+            $path,
+            $defaults,
+            $requirements,
+            ['utf8' => true],
+            $page->domain,
+            null // TODO should we match SSL only if enabled in root? => $page->rootUseSSL ? 'https' : null
+        );
+
+        $this->addRoutesForRootPage($page, $routes);
+    }
+
+    private function addRoutesForRootPage(PageModel $page, array &$routes): void
+    {
+        if ('root' !== $page->type && 'index' !== $page->alias && '/' !== $page->alias) {
+            return;
+        }
+
+        $page->loadDetails();
+
+        $path = '/';
+        $requirements = [];
+        $defaults = $this->getRouteDefaults($page);
+        $table = $this->pageAdapter->getTable();
+
+        if ($this->configAdapter->get('addLanguageToUrl')) {
+            $path = '/{_locale}'.$path;
+            $requirements['_locale'] = $page->rootLanguage;
+        }
+
+        $routes[$table.'.'.$page->id.'.root'] = new Route(
+            $path,
+            $defaults,
+            $requirements,
+            [],
+            $page->domain,
+            null // TODO should we match SSL only if enabled in root? => $page->rootUseSSL ? 'https' : null
+        );
+
+        if ($this->configAdapter->get('addLanguageToUrl') && $page->rootIsFallback) {
+            if (!$this->configAdapter->get('doNotRedirectEmpty')) {
+                $defaults['_controller'] = 'Symfony\Bundle\FrameworkBundle\Controller\RedirectController::urlRedirectAction';
+                $defaults['path'] = '/'.$page->language.'/';
+                $defaults['permanent'] = true;
+            }
+
+            $routes[$table.'.'.$page->id.'.fallback'] = new Route(
+                '/',
+                $defaults,
+                [],
+                [],
+                $page->domain,
+                null // TODO should we match SSL only if enabled in root? => $page->rootUseSSL ? 'https' : null
+            );
+        }
+    }
+
+    private function getRouteDefaults(PageModel $page): array
+    {
+        return [
+            '_token_check' => true,
+            '_controller' => 'Contao\FrontendIndex::renderPage',
+            '_scope' => ContaoCoreBundle::SCOPE_FRONTEND,
+            '_locale' => $page->rootLanguage,
+            'pageModel' => $page,
+        ];
+    }
+
+    private function getPageIdsFromNames(array $names)
+    {
+        $ids = [];
+        $table = $this->pageAdapter->getTable();
+
+        foreach ($names as $name) {
+            if (0 !== strncmp($name, $table.'.', 8)) {
+                continue;
+            }
+
+            [, $id] = explode('.', $name);
+
+            $ids[] = $id;
```

Comment:
We should type-cast to `(int)` here.

Reply:
Implemented in 847914e2ed2909575f53e21e66a8e9f42ed40004